sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,232 @@
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training vector, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            Target values.
+
+        sample_weight : array-like, shape (n_samples,) optional
+            Array of weights that are assigned to individual samples.
+            If not provided, then each sample is given unit weight.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, y = check_X_y(X, y, multi_output=True, y_numeric=True)
+
+        # transformers are designed to modify X which is a 2d dimensional, we
+        # need to modify y accordingly.
+        if y.ndim == 1:
+            y_2d = y.reshape(-1, 1)
+        else:
+            y_2d = y
+        self._fit_transformer(y_2d)
+
+        if self.regressor is None:
+            from ..linear_model import LinearRegression
+            self.regressor_ = LinearRegression()
+        else:
+            if not is_regressor(self.regressor):
+                raise TypeError("The regressor {} is of type {}. Provide"
+                                " a regressor instead.".format(
+                                    self.regressor.__class__.__name__,
+                                    self.regressor._estimator_type))
+            self.regressor_ = clone(self.regressor)
+
+        # transform y and convert back to 1d array if needed
+        y_trans = self.transformer_.fit_transform(y_2d)
+        # FIXME: a FunctionTransformer can return a 1D array even when validate
+        # is set to True. Therefore, we need to check the number of dimension
+        # first.
+        if y_trans.ndim == 2 and y_trans.shape[1] == 1:
+            y_trans = y_trans.squeeze(axis=1)
+        if sample_weight is None:
+            self.regressor_.fit(X, y_trans)
+        else:
+            self.regressor_.fit(X, y_trans, sample_weight=sample_weight)
+
+        return self
+
+    def predict(self, X):
+        """Predict using the base regressor, applying inverse.
+
+        The regressor is used to predict and the ``inverse_func`` or
+        ``inverse_transform`` is applied before returning the prediction.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Samples.
+
+        Returns
+        -------
+        y_hat : array, shape = (n_samples,)
+            Predicted values.
+
+        """
+        check_is_fitted(self, "regressor_")
+        pred = self.regressor_.predict(X)
+        if pred.ndim == 1:
+            pred_trans = self.transformer_.inverse_transform(
+                pred.reshape(-1, 1))
+        else:
+            pred_trans = self.transformer_.inverse_transform(pred)
+        if pred_trans.ndim == 2 and pred_trans.shape[1] == 1:
```

Comment:
I think the shape should match the input rather than always squeezing... but I'm not sure. What does, say, decision tree do with predicting a column array?

Reply:
```python
import numpy as np
from sklearn.tree import DecisionTreeClassifier
X = np.random.random((100, 1))
y = np.array([0] * 50 + [1] * 50)
clf = DecisionTreeClassifier()
clf.fit(X, y).predict(X)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1])
```

So I think that the current squeezing is fine.

Reply:
Apparently we squeeze 1d outputs (for classifiers):

```
>>> DecisionTreeClassifier().fit(X, y.reshape(-1, 1)).predict(X)
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1])
```

for DT regressors as well:

```
>>> DecisionTreeRegressor().fit(X, y.reshape(-1, 1)).predict(X)
array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
```

Reply:
This is behavior is not consistent throughout the library though:

```python
>>> X, y = np.random.randn(10, 10), np.random.randn(10, 1)
>>> Ridge().fit(X, y).predict(X).shape
(10, 1)
```

Maybe we should not squeeze by default and preserve the original shapes for 1d column target data.

Reply:
Maybe we should store the shape of the training data and only squeeze if `y.ndim == 1` in the original `y` provided by the users.

Reply:
Maybe @amueller has an opinion on this matter :)

Reply:
`MLPRegressor` has an even different behavior: warn at fit and squeeze at predict.

```python
>>> import numpy as np
>>> from sklearn.neural_network import MLPRegressor
>>> X, y = np.random.randn(10, 10), np.random.randn(10, 2)
>>> MLPRegressor().fit(X, y).predict(X).shape
(10, 2)
>>> X, y = np.random.randn(10, 10), np.random.randn(10, 1)
>>> MLPRegressor().fit(X, y).predict(X).shape
/volatile/ogrisel/code/scikit-learn/sklearn/neural_network/multilayer_perceptron.py:1306: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
(10,)
>>> X, y = np.random.randn(10, 10), np.random.randn(10)
>>> MLPRegressor().fit(X, y).predict(X).shape
(10,)
```

Reply:
I thought we don't usually squeeze. I'm surprised the trees do. It probably depends on whether it's a classifier, supports multi-label, multi-output multi-class or a regressor.

Reply:
Something like that. I suspect that our convention is: "predict an array that is semantically equivalent to training y, and be consistent across versions for an individual estimator"

Which basically means either choice is fine here... Do we think one would be more/less surprising than the other here? I think given that we have user functions intervening and we make constraints on the input and output of those functions, it might be least surprising for us to always output a shape like the training `y`.